File: lecture/examples/chap-02/nested_list.c
/* ------------------------------------------------------------------------ */
/* C99 implementation of a nested list to represent symbolic expressions    */
/*                                                                          */
/* Introduction to Program Synthesis Lecture (Chapter 02 - Foundations)     */
/*                                                                          */
/* Author: Roman Kalkreuth                                                  */
/* ------------------------------------------------------------------------ */

#include <stdio.h>
#include <stdlib.h> 
#include <string.h>
#include <stdbool.h> 

struct Node* create_list(char* symbols, int *x, bool init_head);
struct Node* create_node(char symbol);
void append_node(struct Node* tail, char symbol);
void nest_node(struct Node* node_ptr, struct Node* sublist, bool append);
void remove_list(struct Node *head);
void traverse(struct Node* head);

struct Node{
    char symbol;
    bool annotated; 
    struct Node* next; 
    struct Node* nesting; 
};

struct Node* create_list(char* symbols, int* x, bool init_head) {
    struct Node* head = NULL;
    struct Node* node_ptr = NULL;
    struct Node* subhead = NULL;

    bool append_nesting = true;

    if(init_head) {
        head = (struct Node*) malloc(sizeof(struct Node));
        head->annotated = false;
        node_ptr = head;
    }

    int n = strlen(symbols);

    while(*x < n) {
       char s = symbols[*x];
       if(s != ' ') {
            if( s == '(') {
                (*x)++;
                subhead = create_list(symbols, x, true);
                if(head == NULL) {
                    head = subhead;
                    node_ptr = head;
                } else {
                    nest_node(node_ptr, subhead, append_nesting);
                    node_ptr = node_ptr->next; 
                }
            } else if (s == ')') {
                break;
            } else {
                if(head->annotated == false) {
                    head->symbol = s;
                    head->annotated = true;
                }
                else {
                    append_node(node_ptr, s);
                    node_ptr = node_ptr->next; 
                }      
            }
        }
        (*x)++;
    }
    return head;
} 

struct Node* create_node(char symbol) {
    struct Node* node = (struct Node*) malloc(sizeof(struct Node));
    node->symbol = symbol;
    node->annotated = true; 
    node->next = NULL;
    return node;
}

void append_node(struct Node* tail, char symbol) {
    struct Node* node = create_node(symbol);
    tail->next = node;
}

void nest_node(struct Node* node_ptr, struct Node* sublist, bool append) {
    if (append){
        struct Node* new_node = create_node('\0');
        node_ptr->next = new_node;
        node_ptr = new_node;
    }  
    node_ptr->nesting = sublist;
}

void remove_list(struct Node *head) {
    struct Node* node_ptr = head;
    struct Node* tmp; 

    while (node_ptr != NULL) {
        if(node_ptr->nesting != NULL){
            remove_list(node_ptr->nesting);
        }
        tmp = node_ptr->next;
        free(node_ptr);
        node_ptr = tmp;
     }  
}

struct Node* tail(struct Node* head) {
    struct Node* node_ptr = head;
    while (node_ptr->next != NULL) {
        node_ptr = node_ptr->next;
     }   
    return node_ptr;
}

void traverse(struct Node* head) {
   struct Node* node_ptr = head;
    while (node_ptr != NULL) {
        printf("%c\n", node_ptr->symbol);
        if(node_ptr->nesting != NULL){
            printf("-- nesting --\n");
            traverse(node_ptr->nesting);
        }
        node_ptr = node_ptr->next;
     }   
}

int main(void) {
    char expr[] = "(+ (* a (+ b c)) (*d (/ e f)))";
    // char expr[] = "(+ a (* b (+ c d)))"; 
    int x = 0; 
    struct Node* head = create_list(expr, &x, false);
    traverse(head);
    remove_list(head);
}

File: lecture/examples/chap-02/nested_list.py
"""
Simple example of a nested list addressed in chapter 2.1
(program representation/tree)

Introduction to Program Synthesis Course
Chair for AI Methodology (AIM)
Faculty of Computer Science, RWTH Aachen University
"""

from __future__ import annotations
from dataclasses import dataclass


@dataclass
class Node:
    """
    Representation of a node that
    is specified with an annotation (symbol) and the
    linkage to neighborhood nodes as well as listings.
    """
    symbol: str = None
    annotated: bool = False
    next: Node = None
    prev: Node = None
    nesting: Node = None


def create_node(symbol: str) -> Node:
    """
    Creates a new node and annotates it with the given symbol.

    :param symbol: symbol to be annotated
    :return: newly created node
    """
    return Node(symbol=symbol, annotated=True)


def append_node(tail: Node, symbol: str):
    """
    Create and append a new node to the tail of the list annotated
    with the given symbol.

    :param tail: tail node of the list
    :param symbol: symbol to be annotated
    """
    node = create_node(symbol=symbol)
    tail.next = node
    node.prev = tail


def nest_node(node_ptr: Node, sublist: Node, append: bool):
    """
    Create a new sublist nesting on a referenced node.

    :param node_ptr: Reference to the node to be nested
    :param sublist: Sublist to be nested
    :param append: append a new node that only serves as a nesting reference
    """
    if append:
        # Annotated the node to append with the termination symbol
        new_node = create_node("\0")
        node_ptr.next = new_node
        node_ptr = new_node
    node_ptr.nesting = sublist


def tail(head: Node) -> Node:
    """
    Iterate the list and return the tail node.

    :param head: head node of the list
    :return: tail node
    """
    node_ptr = head
    while node_ptr.next is not None:
        node_ptr = node_ptr.next
    return node_ptr


def traverse(head: Node, level=0):
    """
    Traverses the list recursively and prints the annotations
    of the visited nodes.

    :param head: head of the list
    :param level: current level, zero by default
    """
    node_ptr = head
    space = '\t' * (level) if level > 0 else ''
    print(space, end="")
    while node_ptr is not None:
        print(node_ptr.symbol, end=" ")
        if node_ptr.nesting is not None:
            print("")
            traverse(node_ptr.nesting, level=level + 1)
        node_ptr = node_ptr.next


def remove_list(head: Node):
    """
    Remove the list recursively.
    :param head: head of the list
    """
    node_ptr = head

    while node_ptr is not None:
        if node_ptr.nesting is not None:
            remove_list(node_ptr.nesting)
        tmp = node_ptr.next
        del node_ptr
        node_ptr = tmp


def create_list(expr: list, init_head: bool, idx: int = 0, append_nesting: bool = True):
    """
    Creates a new nested list recursively based on the given (symbolic) expression.

    :param expr: symbolic expression
    :param init_head: option to init the head node
    :param idx: current node index for the recursion
    :param append_nesting: append option for nestings
    """
    head = None
    node_ptr = None

    if init_head:
        head = Node()
        head.annotated = False
        node_ptr = head


    # Iterate over the expression
    n = len(expr)
    while idx < n:
        s = expr[idx]
        if s != ' ':
            # Identify the start of a new sublist
            if s == '(':
                idx += 1
                subhead, idx = create_list(expr, init_head=True, idx=idx)
                # Initialize the global header, if this has not already been done
                if head is None:
                    head = subhead
                    node_ptr = head
                # Otherwise nest the head of the created sublist
                else:
                    nest_node(node_ptr, subhead, append_nesting)
                    node_ptr = node_ptr.next
            # Leave the nesting recursion level
            elif s == ')':
                break
            else:
                if not head.annotated:
                    head.symbol = s
                    head.annotated = True
                else:
                    append_node(node_ptr, s)
                    node_ptr = node_ptr.next
        idx += 1
    return head, idx


def main():
    expr = "(+ (* a (+ b c)) (* d (/ e f)))"
    symbols = list(expr)
    head, idx = create_list(symbols, False)
    traverse(head)


if __name__ == "__main__":
    main()
